### PR TITLE
matrix-synapse: 0.27.2 -> 0.27.4

### PR DESCRIPTION
**Pull request by florianjacob**
###### Motivation for this change
together with required dependency version bumps.

Mainly brings performance improvements and some bugfixes.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by florianjacob** [comment]
Maintainers: @Ralith @roblabla

**Comment by Ralith** [comment]
LGTM, thanks!

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -2558,11 +2558,11 @@ in {
   fritzconnection = callPackage ../development/python-modules/fritzconnection { };
 
   frozendict = buildPythonPackage rec {
-    name = "frozendict-0.5";
+    name = "frozendict-1.2";
 
     src = pkgs.fetchurl {
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]: Please use `fetchPypi` instead.

File: pkgs/top-level/python-packages.nix [hunk 2]
```diff
@@ -16640,16 +16640,16 @@ EOF
 
   canonicaljson = buildPythonPackage rec {
     name = "canonicaljson-${version}";
-    version = "1.0.0";
+    version = "1.1.3";
 
     src = pkgs.fetchgit {
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]: `fetchFromGitHub`, or even better - if possible - `fetchPypi`

**Review by dotlambda — CHANGES_REQUESTED** [review]
Would you mind moving `frozendict` and `canonicaljson` out of `python-packages.nix`?

**Comment by florianjacob** [comment]
@dotlambda Moving out of `pythonPackages`, like `flake8 = callPackage ../development/python-modules/flake8 { };`?

File: pkgs/top-level/python-packages.nix (hunk 2 above)
Comment (florianjacob) [reply pkgs/top-level/python-packages.nix]: should work with `fetchPypi`, did not know that existed, on it.

**Comment by dotlambda** [comment]
Not out of `pythonPackages`, but yes, something like
```
canonicaljson = callPackage ../development/python-modules/canonicaljson.nix { };
```

**Comment by florianjacob** [comment]
@dotlambda done!

By the way, is there something like `nix-prefetch-pypi`, or what is the intended way to update hashes when fetching from pypi?

**Comment by dotlambda** [comment]
On the new PyPI site (pypi.org), the hashes are right there were you can download the tarball.

**Comment by florianjacob** [comment]
@dotlambda I see, I thought those would not work for some reason like it's the case with fetchurl.

**Comment by florianjacob** [comment]
@peterhoeg applied!

**Comment by dotlambda** [comment]
@GrahamcOfBorg build python2.pkgs.canonicaljson python3.pkgs.canonicaljson python2.pkgs.frozendict python3.pkgs.frozendict

**Review by dotlambda — APPROVED** [review]

**Comment by florianjacob** [comment]
@dotlambda right, frozendict checks are disabled now.

**Comment by peterhoeg** [comment]
As you're on NixOS, please run the tests with ```nix build -f ./nixos/tests/matrix-synapse.nix```

**Comment by florianjacob** [comment]
@peterhoeg I tried that upfront when I was also testing matrix-synapse manually, but when I execute the tests I get trapped in a strange endless loop that I'm failing to debug, repeatedly trying to start the synapse service all over the place.

As I could not find problems during manual testing and the tests seem to work for @GrahamcOfBorg, I assume this is a local problem. (Might be because my NixOS host is a VM, and then having VMs inside VMs is causing problems, or so? :thinking:)

But as the test success/fail score is 1:1 between me and @GrahamcOfBorg, could anybody else try to reproduce whether the tests work or fail, so that we can select a winner?

**Review by dotlambda — APPROVED** [review]
Tests run successfully for me too. I do however not use Matrix so I did not test beyond the NixOS test.

**Review by FRidh — APPROVED** [review]

**Comment by florianjacob** [comment]
Thanks to everyone, learned alot about python packaging. :smile:

**Pull request merged**